oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/AsyncIndexUpdate.java

```diff
@@ -574,25 +575,25 @@ private void runWhenPermitted() {
             return;
         }
 
-        String checkpointToRelease = afterCheckpoint;
+        AtomicReference<String> checkpointToReleaseRef = new AtomicReference<>(afterCheckpoint);
         boolean updatePostRunStatus = false;
         try {
             String newThreadName = "async-index-update-" + name;
             log.trace("Switching thread name to {}", newThreadName);
             threadNameChanged = true;
             Thread.currentThread().setName(newThreadName);
             updatePostRunStatus = updateIndex(before, beforeCheckpoint, after,
-                    afterCheckpoint, afterTime, callback);
+                    afterCheckpoint, afterTime, callback, checkpointToReleaseRef);
 
             // the update succeeded, i.e. it no longer fails
             if (indexStats.didLastIndexingCycleFailed()) {
                 indexStats.fixed();
             }
 
-            // the update succeeded, so we can release the earlier checkpoint
+            // the update succeeded, so we are sure we can release the earlier checkpoint -
             // otherwise the new checkpoint associated with the failed update
-            // will get released in the finally block
-            checkpointToRelease = beforeCheckpoint;
+            // may still get released in the finally block (depending on where the index update failed)
+            checkpointToReleaseRef.set(beforeCheckpoint);
```

Comment:
I'm wondering if this is really necessary - as this is needed for the successful case only - in which case it is already done in [updateIndex() here](https://github.com/oak-indexing/jackrabbit-oak/pull/407/files#diff-2a8cb4a6afd74366e0fc69967d49ed18e9093e6f7e2e1d451f0637968a55fc8fR835)

Reply:
You are right, I don't think it is needed. I thought it can't hurt to have it here as well, in case the method "forgets" to set it (which might be possible in a custom implementation)